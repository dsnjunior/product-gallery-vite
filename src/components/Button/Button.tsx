import clsx from 'clsx';
import { Typography } from '../Typography';

export type ButtonProps = {
  variant?: 'filled' | 'outlined';
  icon?: React.ReactNode;
  iconPosition?: 'start' | 'end' | 'icon-only';
  disabled?: boolean;
  children?: React.ReactNode;
  href?: string;
  otherDomain?: boolean;
  className?: string;
  onClick?: () => void;
};

export function Button({
  variant = 'filled',
  icon,
  iconPosition = 'start',
  disabled,
  children,
  href,
  otherDomain,
  className,
  onClick,
}: ButtonProps) {
  const outlined = variant === 'outlined';
  const iconStart = iconPosition === 'start';
  const iconEnd = iconPosition === 'end';
  const iconOnly = iconPosition === 'icon-only';

  const Tag = href ? 'a' : 'button';

  return (
    <Tag
      className={clsx(
        'flex items-center justify-center',
        'text-emerald-500 disabled:text-gray-300',
        'py-2',
        'rounded-full bg-current border-current border shadow-border transition-colors',
        {
          ['bg-transparent hover:bg-neutral-50']: outlined,
          ['hover:text-emerald-600']: !outlined,
          ['px-2']: iconOnly,
          ['px-6']: !iconOnly,
        },
        className,
      )}
      disabled={disabled}
      onClick={onClick}
      {...(href && { href })}
      {...(otherDomain && { target: '_blank', rel: 'noopener noreferrer' })}
    >
      <span
        className={clsx('flex items-center shrink-0', {
          ['text-neutral-50']: !outlined,
          ['flex-row-reverse']: iconEnd,
        })}
      >
        {icon && <span className={clsx('text-[1.125rem]', { ['ml-1']: iconEnd, ['mr-1']: iconStart })}>{icon}</span>}

        <Typography variant="l100" weight="700" component="span" className={clsx({ ['sr-only']: iconOnly })}>
          {children}
        </Typography>
      </span>
    </Tag>
  );
}
