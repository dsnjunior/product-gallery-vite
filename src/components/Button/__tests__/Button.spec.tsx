import { render, screen } from '@testing-library/react';
import { describe, it } from 'vitest';

import { Button } from '../Button';

describe('Button.tsx', () => {
  it('should match snapshot', async () => {
    render(<Button>Test</Button>);

    expect(await screen.findByRole('button')).toMatchSnapshot();
  });

  it('should render children', async () => {
    render(<Button>Test</Button>);

    expect(await screen.findByText('Test')).toBeInTheDocument();
  });

  it('should render outlined variant', async () => {
    render(<Button variant="outlined">Test</Button>);

    expect(await screen.findByRole('button')).toHaveClass('bg-transparent hover:bg-neutral-50');
  });

  it('should render icon', async () => {
    render(<Button icon={<span data-testid="icon">Icon</span>}>Test</Button>);

    const icon = await screen.findByTestId('icon');
    expect(icon).toBeInTheDocument();
    expect(icon.parentElement).toHaveClass('mr-1');
  });

  it('should render icon at end', async () => {
    render(
      <Button icon={<span data-testid="icon">Icon</span>} iconPosition="end">
        Test
      </Button>,
    );

    const button = await screen.findByRole('button');
    expect(button.firstChild).toHaveClass('flex-row-reverse');
    const icon = await screen.findByTestId('icon');
    expect(icon.parentElement).toHaveClass('ml-1');
  });

  it('should render icon only', async () => {
    render(
      <Button icon={<span data-testid="icon">Icon</span>} iconPosition="icon-only">
        Test
      </Button>,
    );

    const button = await screen.findByRole('button');
    expect(button).toHaveClass('px-2');
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    expect(button.firstChild!.lastChild).toHaveClass('sr-only');
  });

  it('should render disabled', async () => {
    render(<Button disabled>Test</Button>);

    const button = await screen.findByRole('button');
    expect(button).toHaveClass('disabled:text-gray-300');
    expect(button).toBeDisabled();
  });
});
