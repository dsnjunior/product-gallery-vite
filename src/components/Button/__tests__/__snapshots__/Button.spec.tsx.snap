// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button.tsx > should match snapshot 1`] = `
<button
  class="text-emerald-500 disabled:text-gray-300 px-6 py-2 rounded-full bg-current border-current border shadow-border transition-colors hover:text-emerald-600"
>
  <span
    class="flex items-center text-neutral-50"
  >
    <span
      class="text-l100 font-700"
    >
      Test
    </span>
  </span>
</button>
`;
